hierarchicalforecast/methods.py

```diff
@@ -895,6 +913,59 @@ def fit_predict(self,
 
     __call__ = fit_predict
 
+@njit(parallel=True, fastmath=True, error_model="numpy")
+def _shrunk_covariance_schaferstrimmer(residuals: np.ndarray, residuals_mean: np.ndarray, residuals_std: np.ndarray, mint_shr_ridge: float):
+    """Shrink empirical covariance according to the following method:
+        Schäfer, Juliane, and Korbinian Strimmer. 
+        ‘A Shrinkage Approach to Large-Scale Covariance Matrix Estimation and 
+        Implications for Functional Genomics’. Statistical Applications in 
+        Genetics and Molecular Biology 4, no. 1 (14 January 2005). 
+        https://doi.org/10.2202/1544-6115.1175.
+
+    :meta private:
+    """
+    n_timeseries = residuals.shape[0]
+    n_samples = residuals.shape[1]
+    # We need the empirical covariance, the off-diagonal sum of the variance of 
+    # the empirical correlation matrix and the off-diagonal sum of the squared 
+    # empirical correlation matrix.
+    emp_cov = np.zeros((n_timeseries, n_timeseries), dtype=np.float64)
+    sum_var_emp_corr = np.float64(0.0)
+    sum_sq_emp_corr = np.float64(-n_timeseries)
+    factor_emp_corr = np.float64(n_samples / (n_samples - 1))
+    factor_var_emp_cor = np.float64(n_samples / (n_samples - 1)**3)
+    epsilon = np.float64(2e-8)
+    for i in prange(n_timeseries):
+        # Calculate standardized residuals
+        X_i = residuals[i] - residuals_mean[i]
+        Xs_i = X_i / (residuals_std[i] + epsilon)
+        Xs_i_mean = np.mean(Xs_i)
+        for j in range(n_timeseries):
+            # Calculate standardized residuals
+            X_j = residuals[j] - residuals_mean[j]
+            Xs_j = X_j / (residuals_std[j] + epsilon)
+            Xs_j_mean = np.mean(Xs_j)
+            # Empirical covariance
+            emp_cov[i, j] = factor_emp_corr * np.mean(X_i * X_j)
+            # Sum off-diagonal variance of empirical correlation
+            w = (Xs_i - Xs_i_mean) * (Xs_j - Xs_j_mean)
+            w_mean = np.mean(w)
+            sum_var_emp_corr += (i != j) * factor_var_emp_cor * np.sum(np.square(w - w_mean))
+            # Sum squared empirical correlation (off-diagonal correction made by initializing 
+            # with -n_timeseries, so (i != j) not necessary here)
+            sum_sq_emp_corr += np.square(factor_emp_corr * w_mean)
```

Comment:
```suggestion
        for j in range(i + 1):
            # Calculate standardized residuals
            X_j = residuals[j] - residuals_mean[j]
            Xs_j = X_j / (residuals_std[j] + epsilon)
            Xs_j_mean = np.mean(Xs_j)
            # Empirical covariance
            emp_cov[i, j] = emp_cov[j, i] = factor_emp_corr * np.mean(X_i * X_j)
            # Sum off-diagonal variance of empirical correlation
            w = (Xs_i - Xs_i_mean) * (Xs_j - Xs_j_mean)
            w_mean = np.mean(w)
            sum_var_emp_corr += 2 * (i != j) * factor_var_emp_cor * np.sum(np.square(w - w_mean))
            # Sum squared empirical correlation (off-diagonal correction made by initializing 
            # with -n_timeseries, so (i != j) not necessary here)
            sum_sq_emp_corr += 2 * np.square(factor_emp_corr * w_mean)
```

You can exploit the symmetry of the covariance matrix to reduce the number of iterations by $`\frac{1}{2}n\left( n + 1 \right)`$ in the serialised region of the parallel loop, as $`\text{cov}\left( y_{i}, \, y_{j} \right) \equiv \text{cov}\left( y_{j}, \, y_{i} \right)`$.

Reply:
Again a great idea, thanks!

Reply:
Numba doesn't complain, and I made a few other changes that improve the wall time, so again thanks!